Shieldgemma2

RyanMullins opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

ArthurZucker committed:
single commit

ArthurZucker committed:
correct config

ArthurZucker committed:
fixup

ArthurZucker committed:
dummy pt

RyanMullins committed:
Use ShieldGemma2Config in conversion script

RyanMullins committed:
Update src/transformers/models/shieldgemma2/configuration_shieldgemma2.py

RyanMullins committed:
Adding shieldgemma2 to models.__init__.py

RyanMullins committed:
Adding ShieldGemma2 to main __init__.py

ArthurZucker commented on docs/source/en/model_doc/shieldgemma2.md:
```diff
@@ -0,0 +1,88 @@
+model = ShieldGemma2ForImageClassification.from_pretrained(model_id, device_map="auto")
+processor = AutoProcessor.from_pretrained(model_id)
+
+image_1 = # An image you have loaded
+image_2 = # An image you have loaded
+image_3 = # An image you have loaded
```
using online url links should work ! 🤗

RyanMullins replied:
Done.

ArthurZucker commented on src/transformers/models/shieldgemma2/modeling_shieldgemma2.py:
```diff
@@ -0,0 +1,176 @@
+    @replace_return_docstrings(
+        output_type=ShieldGemma2ImageClassifierOutputWithNoAttention, config_class=_CONFIG_FOR_DOC
+    )
+    def forward(self, **kwargs) -> ShieldGemma2ImageClassifierOutputWithNoAttention:
```
let's just use the explicit kwargs here !

ArthurZucker commented on src/transformers/models/shieldgemma2/processing_shieldgemma2.py:
```diff
@@ -0,0 +1,120 @@
+
+logger = logging.get_logger(__name__)
+
+SHIELDGEMMA2_POLICIES: Mapping[str, str] = {
+    "dangerous": (
+        "No Dangerous Content: The image shall not contain content that"
+        " facilitates or encourages activities that could cause real-world harm"
+        " (e.g., building firearms and explosive devices, promotion of"
+        " terrorism, instructions for suicide)."
+    ),
+    "sexual": (
+        "No Sexually Explicit content: The image shall not contain content that"
+        " depicts explicit or graphic sexual acts (e.g., pornography, erotic"
+        " nudity, depictions of rape or sexual assault)."
+    ),
+    "violence": (
+        "No Violence/Gore content: The image shall not contain content that"
+        " depicts shocking, sensational, or gratuitous violence (e.g.,"
+        " excessive blood and gore, gratuitous violence against animals,"
+        " extreme injury or moment of death)."
+    ),
+}
```
That is probably something we want to be on the hub / configurable and serializable!

ArthurZucker commented on src/transformers/models/shieldgemma2/processing_shieldgemma2.py:
```diff
@@ -0,0 +1,120 @@
+        kwargs["padding"] = True
+        kwargs["padding_side"] = "left"
+
+        policies = kwargs.get("policies", list(SHIELDGEMMA2_POLICIES.keys()))
+        policy_definitions = {
+            **SHIELDGEMMA2_POLICIES,
+            **kwargs.get("custom_policies", {}),
+        }
```
this makes it hard to not use the `SHIELDGEMMA2_POLICIES`

ArthurZucker commented on src/transformers/models/shieldgemma2/processing_shieldgemma2.py:
```diff
@@ -0,0 +1,120 @@
+        tokenizer_input_names = self.tokenizer.model_input_names + ["token_type_ids"]
+        image_processor_input_names = self.image_processor.model_input_names
+        return list(dict.fromkeys(tokenizer_input_names + image_processor_input_names))
+
```
let's make sure policies are serialized in a small `test_shieldgemma2_processor` !

ArthurZucker reviewed (approved):
Thanks, small nits but good to go otherwise!

ain-soph committed:
Update shieldgemma2.md

ain-soph commented:
@RyanMullins Submit a PR for this branch as bug fixes.
https://github.com/RyanMullins/transformers/pull/2
- Fix a bug in readme example snippet (return_tensors="pt" for Processor call)
- Use public image link as example (from https://huggingface.co/google/shieldgemma-2-4b-it)
- Fix typo

ain-soph committed:
Update shieldgemma2.md

ain-soph commented:
There is still an logging issue that outputs incorrect info:

```python3
from PIL import Image
import requests
from transformers import AutoProcessor, ShieldGemma2ForImageClassification

model_id = "google/shieldgemma-2-4b-it"
model = ShieldGemma2ForImageClassification.from_pretrained(model_id, device_map="auto")
processor = AutoProcessor.from_pretrained(model_id)

url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg"
image = Image.open(requests.get(url, stream=True).raw)

custom_policies = {
    "key_a": "descrition_a",
    "key_b": "descrition_b",
}

inputs = processor(
    images=[image],
    custom_policies=custom_policies,
    policies=["dangerous", "key_b"],
    return_tensors="pt",
).to(model.device)

output = model(**inputs)
print(output.probabilities)

```

```
$ python test.py
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.12s/it]
Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.50, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.
Keyword argument `custom_policies` is not a valid argument for this processor and will be ignored.
Keyword argument `policies` is not a valid argument for this processor and will be ignored.
tensor([[5.3613e-11, 1.0000e+00],
        [3.7379e-09, 1.0000e+00]], device='cuda:1', grad_fn=<ToCopyBackward0>)
```

As you can see above, the argument `custom_policies` and `policies` are already functioning (the output tensor is only 2 classes now), but still get
```
Keyword argument `custom_policies` is not a valid argument for this processor and will be ignored.
Keyword argument `policies` is not a valid argument for this processor and will be ignored.
```

ain-soph commented:
And currently it's confusing about the output tensor.
- What's the meaning of each dimension for the output tensor shape?
- What's the meaning of each number (especially the `1.0` in second dimension, see example output above) and which harmful category is it mapped to?
It would be nice if we have something for explanation in the readme / API doc.

RyanMullins committed:
Adding tests. Addressing review feedback.

RyanMullins committed:
Minor docs update

RyanMullins committed:
Fixing code quality feedback from CI

RyanMullins commented:
@ain-soph thanks for the contribs. For some reason I can't reply to your comments, but I've added docstrings to address [your questions](https://github.com/huggingface/transformers/pull/36678#issuecomment-2730504872) about the shape of the output tensors, and I added an explicit `ShieldGemma2ProcessorKwargs` to address the [logs you saw](https://github.com/huggingface/transformers/pull/36678#issuecomment-2730493230).

ghunkins commented:
When running the default example, an `IndexError` is raised. Passing in custom_policies fixes that.

#### Code

```python
from PIL import Image
import requests
from transformers import AutoProcessor, ShieldGemma2ForImageClassification

model_id = "google/shieldgemma-2-4b-it"
model = ShieldGemma2ForImageClassification.from_pretrained(model_id, device_map="auto", token="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",)
processor = AutoProcessor.from_pretrained(model_id, token="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",)

url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg"
image = Image.open(requests.get(url, stream=True).raw)

inputs = processor(images=[image], return_tensors="pt").to(model.device)

output = model(**inputs)
print(output.probabilities)
```

#### Stack Trace

```python
Loading checkpoint shards: 100%
 2/2 [00:06<00:00,  2.86s/it]
WARNING:accelerate.big_modeling:Some parameters are on the meta device because they were offloaded to the cpu.
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
[<ipython-input-26-290947fc3325>](https://localhost:8080/#) in <cell line: 0>()
     10 image = Image.open(requests.get(url, stream=True).raw)
     11 
---> 12 inputs = processor(images=[image], return_tensors="pt").to(model.device)
     13 
     14 output = model(**inputs)

1 frames
[/usr/local/lib/python3.11/dist-packages/transformers/processing_utils.py](https://localhost:8080/#) in apply_chat_template(self, conversation, chat_template, **kwargs)
   1326 
   1327         if isinstance(conversation, (list, tuple)) and (
-> 1328             isinstance(conversation[0], (list, tuple)) or hasattr(conversation[0], "content")
   1329         ):
   1330             is_batched = True

IndexError: list index out of range
```

ghunkins commented:
When running the default example, an IndexError is raised. Passing in custom_policies fixes that.

#### Example

```python
from PIL import Image
import requests
from transformers import AutoProcessor, ShieldGemma2ForImageClassification

model_id = "google/shieldgemma-2-4b-it"
model = ShieldGemma2ForImageClassification.from_pretrained(model_id, device_map="auto")
processor = AutoProcessor.from_pretrained(model_id)

url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg"
image = Image.open(requests.get(url, stream=True).raw)

inputs = processor(images=[image], return_tensors="pt").to(model.device)

output = model(**inputs)
print(output.probabilities)
```

#### Stack Trace
```python
Loading checkpoint shards: 100%
 2/2 [00:06<00:00,  2.86s/it]
WARNING:accelerate.big_modeling:Some parameters are on the meta device because they were offloaded to the cpu.
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
[<ipython-input-26-290947fc3325>](https://localhost:8080/#) in <cell line: 0>()
     10 image = Image.open(requests.get(url, stream=True).raw)
     11 
---> 12 inputs = processor(images=[image], return_tensors="pt").to(model.device)
     13 
     14 output = model(**inputs)

1 frames
[/usr/local/lib/python3.11/dist-packages/transformers/processing_utils.py](https://localhost:8080/#) in apply_chat_template(self, conversation, chat_template, **kwargs)
   1326 
   1327         if isinstance(conversation, (list, tuple)) and (
-> 1328             isinstance(conversation[0], (list, tuple)) or hasattr(conversation[0], "content")
   1329         ):
   1330             is_batched = True

IndexError: list index out of range
```

ArthurZucker commented on src/transformers/models/shieldgemma2/processing_shieldgemma2.py:
```diff
@@ -0,0 +1,182 @@
+                the base policies ShieldGemma was trained on.
+        """
+        super().__init__(image_processor, tokenizer, chat_template, image_seq_length, **kwargs)
+        if policy_definitions:
+            self.policy_definitions = policy_definitions
+        else:
+            self.policy_definitions = {}
```
```suggestion
        if policy_definitions is None:
            self.policy_definitions = DEFAULT_GEMMA_POLICY
        else:
            self.policy_definitions = policy_definitions
```
imo this is th ebst of all worlds!

RyanMullins replied:
Done.

RyanMullins replied:
Done.

ArthurZucker reviewed (approved):
LGTM! THanks for iterating!

RyanMullins committed:
Fixing empty messages bug reported by ghunkins

RyanMullins commented:
@ghunkins thanks for the bug report! 

The problem was that the files on the Hub were out of date and the processor.json was missing the `policy_definitions` property. I was using updated files locally, which is why I didn't see this until I repro'd your example in Colab. The files are now [updated](https://huggingface.co/google/shieldgemma-2-4b-it/blob/main/processor_config.json) on the Hub. 

@ArthurZucker accidentally [suggested a local solution](https://github.com/huggingface/transformers/pull/36678#discussion_r2003461909) to the problem by providing a default set of policies. So this should be fully fixed now and ready for merging.

ghunkins commented:
Brilliant, thanks for the amazing work on this @RyanMullins !

ArthurZucker reviewed (commented):
@RyanMullins let's just fix cis! 🤗

ArthurZucker commented:
Do you need help on my side for the last remaining tests?

ain-soph commented:
nit: I just tested and there is the same warning for `policy_definitions`.
```Some kwargs in processor config are unused and will not have any effect: policy_definitions.```
